patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/UIModBrowser.cs

```diff
@@ -184,103 +236,139 @@ public void BackClick(UIMouseEvent evt, UIElement listeningElement)
 	}
 
 	public override void OnActivate()
 	{
+		base.OnActivate();
 		Main.clrInput();
-		if (!Loading && _items.Count <= 0) {
-			if (WorkshopHelper.QueryHelper.Items.Count == 0)
-				PopulateModBrowser(uiOnly: false);
-			else
-				PopulateModBrowser(uiOnly: true);
+		if (_firstLoad) {
+			SocialBackend.Initialize(); // Note this is currently synchronous
+			PopulateModBrowser();
 		}
+
+		// Check for mods to update
+		// @NOTE: Now it's done only once on load
+		CheckIfAnyModUpdateIsAvailable();
+
+		DebounceTimer = null;
 	}
 
-	internal void PopulateModBrowser(bool uiOnly)
+	private void CbLocalModsChanged(HashSet<string> modSlugs)
 	{
-		// Initialize
-		Loading = true;
-		SpecialModPackFilter = null;
-		SpecialModPackFilterTitle = null;
-		_reloadButton.SetText(Language.GetText("tModLoader.MBGettingData"));
-		_backgroundElement.Append(_loaderElement);
-		SetHeading(Language.GetText("tModLoader.MenuModBrowser"));
+		// Can be called outside main thread
+		lock (modSlugsToUpdateInstallInfo) {
+			modSlugsToUpdateInstallInfo.UnionWith(modSlugs);
+		}
+		// Updates the 'Update All' button using cached data from GetInstalledModDownloadedItems
+		CheckIfAnyModUpdateIsAvailable();
+	}
 
-		// Remove old data
-		_items.Clear();
-		
-		// Asynchronous load the Mod Browser
-		Task.Run(() => {
-			InnerPopulateModBrowser(uiOnly: uiOnly);
-			Loading = false;
-			_reloadButton.SetText(Language.GetText("tModLoader.MBReloadBrowser"));
-		});
+	public override void OnDeactivate()
+	{
+		DebounceTimer = null;
+		base.OnDeactivate();
 	}
 
-	internal bool InnerPopulateModBrowser(bool uiOnly)
+	public override void Update(GameTime gameTime)
 	{
-		if (!uiOnly && !WorkshopHelper.QueryHelper.QueryWorkshop())
-			return false;
+		base.Update(gameTime);
+
+		if (_updateAvailable) {
+			_updateAvailable = false;
+			Interface.updateMessage.SetMessage(_updateText);
+			Interface.updateMessage.SetGotoMenu(Interface.modBrowserID);
+			Interface.updateMessage.SetURL(_updateUrl);
+			Interface.updateMessage.SetAutoUpdateURL(_autoUpdateUrl);
+			Main.menuMode = Interface.updateMessageID;
+		}
+
+		lock (modSlugsToUpdateInstallInfo) {
+			foreach (var item in ModList.ReceivedItems.Where(
+				d => modSlugsToUpdateInstallInfo.Contains(d.ModDownload.ModName)
+			)) {
+				item.ModDownload.UpdateInstallState();
+				item.UpdateInstallInfo();
+			}
+			// @TODO: Shouldn't only delete processed slugs?
+			modSlugsToUpdateInstallInfo.Clear();
+		}
+
+		if (
+			(DebounceTimer is not null) &&
+			(DebounceTimer.ElapsedMilliseconds >= DEBOUNCE_MS)
+		) {
+			// No need to count more
+			DebounceTimer.Stop();
+			DebounceTimer = null;
+		}
 
-		foreach (var item in WorkshopHelper.QueryHelper.Items) {
-			_items.Add(new UIModDownloadItem(item));
+		if (UpdateNeeded) {
+			// Debounce logic
+			if (DebounceTimer is null) {
+				UpdateNeeded = false;
+				PopulateModBrowser();
+				DebounceTimer = new();
+				DebounceTimer.Start();
+			}
```

Comment:
No need for complicated `Stopwatch` implementation. A simple `DateTime _lastUpdate` combined with:
```cs
if (UpdateNeeded && DateTime.Now - _lastUpdate > MinTimeBetweenUpdates)
    PopulateModBrowser();
```